verl/trainer/ppo/ray_trainer.py

```diff
@@ -643,7 +645,21 @@ def _validate(self):
             test_batch = test_batch.union(test_output_gen_batch)
 
             # evaluate using reward_function
-            reward_tensor = self.val_reward_fn(test_batch)
+            reward_result = self.val_reward_fn(test_batch)
+
+            # Handle both scalar and dictionary returns
+            if isinstance(reward_result, dict):
+                reward_tensor = reward_result['reward_tensor']
+                if 'extra_info' in reward_result:
+                    if reward_extra_info_dict is None:
+                        reward_extra_info_dict = {}
+                    for key, extra_reward in reward_result['extra_info'].items():
+                        if key not in reward_extra_info_dict:
```

Comment:
i think this can be simplified by `defaultdict(list)` and you can all append directly